File: LC/Python/Medium/2958.py
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        ans = 0
        window_start = 0
        freq = defaultdict(int)        
        
        for window_end in range(len(nums)):
            freq[nums[window_end]] += 1

            while freq[nums[window_end]] > k:
                freq[nums[window_start]] -= 1
                window_start += 1
            
            ans = max(ans, window_end - window_start + 1)
        
        return ans

File: LC/C++/Medium/2958.cpp
class Solution {
public:
    int maxSubarrayLength(vector<int>& nums, int k) {
        size_t ans = 0;
        size_t l = 0;
        std::unordered_map<size_t, int> seen{};

        for (size_t r = 0; r < nums.size(); ++r) {
            seen[nums[r]]++;

            while (seen[nums[r]] > k) {
                seen[nums[l]]--;
                l++;
            }

            ans = std::max(ans, r - l + 1);
        }

        return ans;
    }
};